internal/pkg/manifest/union.go

```diff
@@ -0,0 +1,152 @@
+	isAdvanced bool
+
+	// Advanced holds the value of Union if IsAdvanced() is true.
+	// If IsAdvanced() is false, this is the zero value of type Advanced.
+	//
+	// Advanced is exported to support mergo. It should not be set
+	// directly. Use NewUnionAdvanced() to create a Union with Advanced set.
+	Advanced Advanced
+}
+
+// BasicToUnion creates a new Union[Basic, Advanced] with the underlying
+// type set to Basic, holding val.
+func BasicToUnion[Basic, Advanced any](val Basic) Union[Basic, Advanced] {
+	return Union[Basic, Advanced]{
+		isBasic: true,
+		Basic:   val,
+	}
+}
+
+// NewUnionAdvanced creates a new Union[Basic, Advanced] with the underlying
+// type set to Advanced, holding val.
+func NewUnionAdvanced[Basic, Advanced any](val Advanced) Union[Basic, Advanced] {
+	return Union[Basic, Advanced]{
+		isAdvanced: true,
+		Advanced:   val,
+	}
+}
+
+// IsBasic returns true if the underlying value of t is type Basic.
+func (t *Union[_, _]) IsBasic() bool {
+	return t.isBasic
+}
+
+// IsAdvanced returns true if the underlying value of t is type Advanced.
+func (t *Union[_, _]) IsAdvanced() bool {
+	return t.isAdvanced
+}
+
+// UnmarshalYAML decodes value into either type Basic or Advanced, and stores that value
+// in t. value is first decoded into type Basic, and t will hold type Basic if:
+//   - There was no error decoding value into type Basic
+//   - Basic.IsZero() returns false OR Basic does not support IsZero()
+//
+// If Basic didn't meet the above criteria, then value is decoded into type Advanced.
+// t will hold type Advanced if Advanced meets the same conditions that were required for type Basic.
+//
+// If value fails to decode into either type, or both types are zero after
+// decoding, t will not hold any value.
+func (t *Union[Basic, Advanced]) UnmarshalYAML(value *yaml.Node) error {
+	// reset struct
+	var simple Basic
+	var advanced Advanced
+	t.isBasic, t.Basic = false, simple
+	t.isAdvanced, t.Advanced = false, advanced
+
+	sErr := value.Decode(&simple)
+	if sErr == nil && !isZeroerAndZero(simple) {
+		t.isBasic, t.Basic = true, simple
+		return nil
+	}
+
+	aErr := value.Decode(&advanced)
+	if aErr == nil && !isZeroerAndZero(advanced) {
+		t.isAdvanced, t.Advanced = true, advanced
+		return nil
+	}
+
+	if sErr != nil {
+		return sErr
+	}
+	return aErr
+}
+
+// isZeroerAndZero returns true if v is a yaml.Zeroer
+// _and_ is zero, as determined by v.IsZero().
+func isZeroerAndZero(v any) bool {
+	z, ok := v.(yaml.IsZeroer)
+	if !ok {
+		return false
+	}
+	return z.IsZero()
+}
+
+// MarshalYAML implements yaml.Marshaler.
+func (t Union[_, _]) MarshalYAML() (interface{}, error) {
+	switch {
+	case t.IsBasic():
+		return t.Basic, nil
+	case t.IsAdvanced():
+		return t.Advanced, nil
+	}
+	return nil, nil
+}
+
+// IsZero returns true if the set value of t implements
+// yaml.IsZeroer and IsZero. It also returns true if
+// neither value for t is set.
+func (t Union[_, _]) IsZero() bool {
```

Comment:
nit: can these be pointer receivers as well so that all receivers are consistent?

Reply:
the three funcs that have value receivers (`MarshalYAML()`, `IsZero()`, and `validate()`) all get called after the type has been cast to an interface, and since `Union` is embedded, not `*Union`, they have to be on the value receiver.